src/com/sap/piper/cm/ChangeManagement.groovy

```diff
@@ -172,30 +172,141 @@ public class ChangeManagement implements Serializable {
+
+        // --> Default config has been adjusted so that no dependencies needs to be installed.
+        // This is the case when an image is used which contains already all dependencies.
+        // In this case we don't invoke npm install and we run the image with the standard user
+        // already, since there is no need for being root. Hence we don't need to switch user also
+        // in the script.
+        boolean noInstall = deployToolDependencies.isEmpty()
+
+        def cmd = ['#!/bin/bash -e']
+
+        if (! noInstall) {
+            cmd << "npm install -g ${deployToolDependencies}"
+            cmd << "su ${osDeployUser}"
+        }
+        cmd << "fiori deploy -c \"${deployConfigFile}\" -- -y"
+
+        // 3.) execute the call in an appropirate docker container (node) and evaluate the return code
+        //     or let the AbortException bubble up.
+        script.withCredentials([script.usernamePassword(
+            credentialsId: credentialsId,
+            passwordVariable: 'password',
+            usernameVariable: 'username')]) {
+
+            // Set username and password for the fiori deploy call. The config file is configured to read the
+            // credentials from the environment (see above in the config file template).
+            // After installing the deploy toolset we switch the user. Since we do not su with option '-l' the
+            // environment variables are preserved.
+            def dockerEnvVars = docker.envVars ?: [:] + [ABAP_USER: script.username, ABAP_PASSWORD: script.password]
+
+            def dockerOptions = docker.options ?: []
+            if (!noInstall) {
+                // when we install globally we need to be root, after preparing that we can su node` in the bash script.
+                dockerOptions += ['-u', '0'] // should only be added if not already present.
+            }
+
+            script.dockerExecute(
+                script: script,
+                dockerImage: docker.image,
+                dockerOptions: dockerOptions,
+                dockerEnvVars: dockerEnvVars,
+                dockerPullImage: docker.pullImage) {
+
+                script.sh script: (cmd as Iterable).join('\n')
+            }
         }
+        // === Dungheap ===
+        // We need to cross check the dependencies between a project and our deployment code. e.g. the fiori toolset
+        // expects the folder containing the app inside a folder 'dist' (hard coded).
+        //
+        // In the meantime the code is not well structed anymore. We started with supporting the cm client only.
+        // Afterwards we added RFC upload support. Now we use node based toolset for the CTS upload. We have now three
+        // different toolset for three ways to perform the upload. The general code flow cannot be explained anymore to
+        // anybody. ==> we should rework that. Makes also a shift to go easier at a later point in time when the code is
+        // well structured.
```

Comment:
Should we create separate steps for the three different toolsets and then call the respective one from this ChangeManagement.groovy file?

Reply:
We started with one scenario (SOLMAN). Afterwards we introduced CTS. And after that RFC for customers with older ABAP versions. Since all deals with uploading content into ABAP system the decision was made to have one step with several modes. Hard to find out the details for that decision making process now. Since we have that state now I would like to suggest to keep that like it is.

Reply:
Ok, so you are suggesting to structure the file differently, or what?

Reply:
> Ok, so you are suggesting to structure the file differently, or what?

yes